project/src/main/java/com/google/sps/data/VisionLibrary.java

```diff
@@ -0,0 +1,330 @@
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the BlobKey is empty. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /** Returns a URL that points to the uploaded file. */
+  public static String getUploadedFileUrl(ImagesService imagesService, BlobKey blobKey) {
+    // In the case that the user did not upload any image, return null.
+    if (blobKey == null) return null;
+    // Attempt to use imagesService API, otherwise serve blob directly through
+    // a servlet.
+    try {
+      ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+      // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+      // path to the image, rather than the path returned by imagesService which contains a host.
+      try {
+        URL url = new URL(imagesService.getServingUrl(options));
+        return url.getPath();
+      } catch (MalformedURLException e) {
+        return imagesService.getServingUrl(options);
+      }
+    } catch (Exception e) {
+      return "/serveBlobstoreImage?blobKey=" + blobKey.getKeyString();
+    }
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
+   */
+  public static byte[] getBlobBytes(BlobstoreService blobstoreService, BlobKey blobKey) throws IOException {
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // if we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+  
+  public static String getNearestColor(Color color) {
+    Color[] constantColors = new Color[] { Color.black, Color.blue, Color.cyan, Color.gray, Color.green, Color.magenta, Color.orange, Color.pink, Color.red, Color.white, Color.yellow };
```

Comment:
It would be good to break this up (maybe in the middle) so it fits within 80 characters.